java-client/src/main/java/org/apache/pegasus/client/PegasusScanner.java

```diff
@@ -88,9 +88,31 @@ public PegasusScanner(
     _needCheckHash = needCheckHash;
     _incomplete = false;
     _fullScan = fullScan;
+    _nextItem = null;
+  }
+
+  public boolean hasNext() throws PException {
+    synchronized (_nextItemLock) {
+      if (_incomplete) {
+        return false;
+      }
+      if (_rpcRunning && !_encounterError) {
+        return !_promises.isEmpty();
+      } else {
+        _nextItem = next();
+        return _nextItem != null;
+      }
+    }
```

Comment:
I think you don't have to make such complex, just like follow seem to be ok:
```java
public boolean hasNext() throws PException {
        _nextItem = next();
        return _nextItem != null;
}
```

`next()` method include enough logic to judge current status and make sure `thread-safe`. 

That is to say `_incomplete`, `_rpcRunning && !_encounterError` should throw exception but return fasle

Reply:
Yes, but if hasNext is called many times individually,  it will cause the iterator to move,  which is not desirable.

Reply:
> but if hasNext is called many times individually, it will cause the iterator to move, which is not desirable
 This case is right, so you may need update it like:
```java
public boolean hasNext() throws PException {
          synchronized (_nextItemLock) {
            if (_nextItem != null) {
                 return true;
            }
        }
        _nextItem = next();
        return _nextItem != null;
}
```